File: server/internal/store/spelling.go
package store

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
)

// Forgiving a typo, without pretending that also fixes a missing name.
//
// A typo (`ardvark`, `betual`) is one or two edits from a real string; a missing
// name (`hard maple`, not in the corpus) is 3–4 edits from anything, and a
// matcher loose enough to reach it would be wrong more often than right. This
// fixes the first; `pipeline/.../spelling.py` is the design and measurements.
//
// Nothing here is on the hot path: `/v1/search` answers as it did, and only when
// its answer is Answer.Weak does Store.Suggest run and the unchanged search run
// again on the corrected string. The reader who pays for the second pass is the
// one who was getting nothing usable anyway.
//
// `spellingKey` is a second implementation of the pipeline's `spelling_key`
// (like `abbreviateBinomial`): the pipeline keys the corpus, this keys the
// query, and a disagreement returns nothing silently, so `spelling_test.go`
// samples the built table and requires this to reproduce the stored key. That
// risk is why the key is fifteen lines rather than Double Metaphone.

// SpellingSchema is the built typo index: one row per distinct word across
// every name the search can match, keyed phonetically.
//
// Absent on any build predating the index, in which case a query that returns
// nothing simply returns nothing — the same answer as before, which is why
// nothing downstream needs a fallback.
type SpellingSchema struct {
	Table string `json:"table"`
	Key   string `json:"key"`
	Word  string `json:"word"`
	// How many distinct names carry the word. The tiebreak between two
	// candidates at equal distance, and the only ordering signal a bare word
	// has. Optional: without it ties fall to length then lexicographic, which
	// is still deterministic.
	N string `json:"n,omitempty"`
}

func (s *Schema) resolveSpelling() {
	t := s.firstTable("spelling")
	if t == "" {
		return
	}
	key, word := s.col(t, "key"), s.col(t, "word")
	if key == "" || word == "" {
		s.Skipped[t] = "no key/word column pair could be resolved"
		return
	}
	s.Spelling = &SpellingSchema{Table: t, Key: key, Word: word, N: s.col(t, "n")}
}

const (
	// A word shorter than this is never corrected: every measured false
	// correction was a short word one edit from another, and this floor takes the
	// false-correction rate on junk from 25.3% to 0.5% while costing nothing
	// (every real misspelling is six characters or longer).
	minCorrectedWord = 6

	// The edit budget, relative to length: one edit is meaningless on twenty
	// characters, two on six is the `hard maple` mistake with a smaller number.
	longWord         = 10
	maxDistanceShort = 1
	maxDistanceLong  = 2

	// A guard on the bucket scan; buckets are tiny, but an unbounded loop over an
	// index this code did not build is a promise nobody checked.
	spellingBucketCap = 4096
)

func distanceCap(word string) int {
	if len(word) < longWord {
		return maxDistanceShort
	}
	return maxDistanceLong
}

// spellingWords splits a query into the words the index holds. Returns nothing
// for any non-ASCII string rather than splitting around it (`aapajärvensis` must
// not become `aapaj` + `rvensis`); the pipeline applies the same test, which
// lets both sides agree without a Unicode normaliser.
func spellingWords(text string) []string {
	for i := range len(text) {
		if text[i] >= 0x80 {
			return nil
		}
	}
	var out []string
	start := -1
	for i := range len(text) + 1 {
		var c byte
		if i < len(text) {
			c = lowerASCII(text[i])
		}
		if i < len(text) && (c >= 'a' && c <= 'z' || c >= '0' && c <= '9') {
			if start < 0 {
				start = i
			}
			continue
		}
		if start >= 0 {
			out = append(out, strings.ToLower(text[start:i]))
			start = -1
		}
	}
	return out
}

func lowerASCII(c byte) byte {
	if c >= 'A' && c <= 'Z' {
		return c + ('a' - 'A')
	}
	return c
}

func isSpellingVowel(c byte) bool {
	switch c {
	case 'a', 'e', 'i', 'o', 'u', 'y':
		return true
	}
	return false
}

// foldPH folds `ph`→`f` for both the key and the distance. It must be in both:
// folding it into the key alone puts `elefant` in `elephant`'s bucket, then the
// distance charges two edits over a cap of one. Not a wider cap — one
// substitution just stops being counted twice.
func foldPH(word string) string { return strings.ReplaceAll(word, "ph", "f") }

// spellingKey is the phonetic key of one lowercased ASCII word: `ph`→`f`,
// drop non-initial `h`, keep the first letter, drop vowels and repeated letters
// (aardvark/ardvark -> ardvrk). Mirrored exactly by `spelling_key` in the
// pipeline; `spelling_test.go` holds the two together.
func spellingKey(word string) string {
	folded := foldPH(word)
	if folded == "" {
		return ""
	}
	var b strings.Builder
	b.Grow(len(folded))
	last := folded[0]
	b.WriteByte(last)
	for i := 1; i < len(folded); i++ {
		c := folded[i]
		// Silent `h`, and only after the first letter — `Homo` keeps its own.
		if c == 'h' || isSpellingVowel(c) || c == last {
			continue
		}
		b.WriteByte(c)
		last = c
	}
	return b.String()
}

// damerau is optimal string alignment distance, abandoning once a whole row
// exceeds the budget. Transpositions count as one edit: `betual`→`betula` is one
// swap, and under plain Levenshtein it is two, tying with `betual`→`betel` (a
// different plant) where the shorter string would win.
func damerau(a, b string, budget int) int {
	la, lb := len(a), len(b)
	if la-lb > budget || lb-la > budget {
		return budget + 1
	}
	prev2 := make([]int, lb+1)
	prev := make([]int, lb+1)
	cur := make([]int, lb+1)
	for j := 0; j <= lb; j++ {
		prev[j] = j
	}
	for i := 1; i <= la; i++ {
		cur[0] = i
		best := cur[0]
		for j := 1; j <= lb; j++ {
			sub := prev[j-1]
			if a[i-1] != b[j-1] {
				sub++
			}
			v := min(prev[j]+1, cur[j-1]+1, sub)
			if i > 1 && j > 1 && a[i-1] == b[j-2] && a[i-2] == b[j-1] {
				v = min(v, prev2[j-2]+1)
			}
			cur[j] = v
			best = min(best, v)
		}
		if best > budget {
			return budget + 1
		}
		prev2, prev, cur = prev, cur, prev2
	}
	return prev[lb]
}

// Suggest returns a better spelling of q, or "" when there is nothing to fix.
//
// Word by word, since with typeahead the misspelling that matters is in the
// leading token (`betual` kills the query before `pendula` is typed), which
// whole-name matching cannot correct. A suggestion, not an answer: the caller
// re-runs the search and reports it only if it produced results.
func (s *Store) Suggest(ctx context.Context, q string) (string, error) {
	sp := s.Schema.Spelling
	if sp == nil {
		return "", nil
	}
	words := spellingWords(q)
	if len(words) == 0 {
		return "", nil
	}
	out := make([]string, len(words))
	changed := false
	for i, w := range words {
		fixed, err := s.suggestWord(ctx, w)
		if err != nil {
			return "", err
		}
		if fixed != "" {
			out[i], changed = fixed, true
		} else {
			out[i] = w
		}
	}
	if !changed {
		return "", nil
	}
	return strings.Join(out, " "), nil
}

// suggestWord is the whole of the ranking: one indexed lookup, then Damerau
// over the handful of words that share the key.
func (s *Store) suggestWord(ctx context.Context, word string) (string, error) {
	if len(word) < minCorrectedWord {
		return "", nil
	}
	sp := s.Schema.Spelling
	n := "0"
	if sp.N != "" {
		n = fmt.Sprintf("%q", sp.N)
	}
	rows, err := s.DB.QueryContext(ctx, fmt.Sprintf(
		`SELECT %q, %s FROM %q WHERE %q = ? LIMIT ?`,
		sp.Word, n, sp.Table, sp.Key), spellingKey(word), spellingBucketCap)
	if err != nil {
		return "", err
	}
	defer rows.Close() //nolint:errcheck

	budget := distanceCap(word)
	// Both sides folded, since the bucket was built on the folded form. See foldPH.
	folded := foldPH(word)
	bestWord := ""
	bestDist, bestUse := budget+1, int64(-1)
	for rows.Next() {
		var cand string
		var use sql.NullInt64
		if err := rows.Scan(&cand, &use); err != nil {
			return "", err
		}
		// A word the corpus holds is registered, not a typo, so never correct it.
		if cand == word {
			return "", nil
		}
		d := damerau(folded, foldPH(cand), budget)
		if d > budget {
			continue
		}
		// Distance, then more widely used, then shorter, then lexicographic — so
		// the answer never depends on SQLite's row order.
		if bestWord == "" || betterSuggestion(d, use.Int64, cand, bestDist, bestUse, bestWord) {
			bestWord, bestDist, bestUse = cand, d, use.Int64
		}
	}
	if err := rows.Err(); err != nil {
		return "", err
	}
	return bestWord, nil
}

func betterSuggestion(d int, use int64, word string, bd int, bu int64, bw string) bool {
	if d != bd {
		return d < bd
	}
	if use != bu {
		return use > bu
	}
	if len(word) != len(bw) {
		return len(word) < len(bw)
	}
	return word < bw
}

File: pipeline/src/concestor_build/spelling.py
"""Forgiving a typo, and refusing to forgive a missing name.

A typo (`ardvark`, `betual`) is one or two edits from a string the corpus holds.
A missing name (`hard maple` for *Acer saccharum*, not in the corpus) is 3–4
edits from anything real, and a matcher loose enough to reach it would be wrong
more often than right. This fixes the first and is designed so no tuning fixes
the second.

# The shape

Correct the query; never relax the matcher. Nothing here runs on the hot path:
`/v1/search` is unchanged, and only when it returns nothing does the server ask
this index for a spelling and re-run the same pipeline. Recall and ranking are
separate:

1. Recall is a phonetic key, an indexed column (the server is pure-Go
   `modernc.org/sqlite` and cannot load `spellfix1`/`editdist3`).
2. Ranking is Damerau-Levenshtein, in Go, over the handful the key returns.

# The key, and why it is not Double Metaphone

The key must be computed identically in Python (corpus) and Go (query), or the
lookup silently returns nothing. Double Metaphone is hundreds of order-dependent
rules — two ports, two chances to disagree invisibly — so the key is fifteen
lines instead:

    lowercase, ASCII only, split on non-alphanumerics; then per word —
    `ph` -> `f`, drop non-initial `h`, keep the first letter, drop vowels
    (including `y`) and collapse runs of the same letter.

    aardvark, ardvark  -> ardvrk    ;    dolphin, dolfin -> dlfn

Plain vowel-dropping scores 16/20 on measured misspellings; `ph`->`f` and silent
`h` take it to 19/20. Every further sound rule was refused for costing precision
(folding `z`/`q` puts the benchmark `zzzqqq` in a 69-candidate bucket).

Non-ASCII words (0.27% of the corpus) are not indexed, which avoids a Unicode
normalisation dependency in Go for a slice nobody misspells.

# Words, not whole names

The unit is the word. Whole-name matching is 7x larger and cannot correct
`betual pendula` at all, since the misspelling is in the leading token. Words are
denser, so the precision cost is short words: every measured false correction was
a single edit on four or five characters, so a word shorter than `MIN_CORRECTED`
is never corrected — taking the false-correction rate from 25.3% to 0.5% while
costing nothing (every real misspelling is six characters or longer).

Other guards: a word the corpus already holds is never a typo; a correction that
yields nothing is not reported; the edit cap is length-relative (one under ten
characters, two at ten or more).

# What this deliberately does not fix

`hard maple` produces no candidates: its key `hrd mpl` does not connect to
`sugar maple`'s `sgr mpl`. That is correct — a phase 6 coverage gap, not a
matcher failure.
"""

from __future__ import annotations

import re
import time
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    import sqlite3
    from collections.abc import Iterable, Iterator

    from .typing_ import Log

SCHEMA = """
DROP TABLE IF EXISTS spelling;

-- One row per distinct word across every name the search can match, keyed by
-- the phonetic key above. `n` is how many distinct names carry the word, and it
-- is the tiebreak between two candidates at the same edit distance — the more
-- widely used spelling wins, which is the only ordering signal a bare word has.
CREATE TABLE spelling (
  key  TEXT NOT NULL,
  word TEXT NOT NULL,
  n    INTEGER NOT NULL
);
"""

INDEXES = """
CREATE INDEX spelling_key ON spelling(key);
"""

VOWELS = frozenset("aeiouy")
_WORD = re.compile(r"[a-z0-9]+")

# A word shorter than this is never corrected, and the floor is where all the
# measured precision lives. See the module docstring.
MIN_CORRECTED = 6

# Nothing shorter than this can be the *target* of a correction either, so there
# is no reason to index it: a query word is at least MIN_CORRECTED long and may
# move by at most MAX_DISTANCE_LONG.
MIN_INDEXED = 4

# The edit budget, relative to the length of the word being corrected. One edit
# is generous on six characters and meaningless on twenty.
LONG_WORD = 10
MAX_DISTANCE_SHORT = 1
MAX_DISTANCE_LONG = 2


def distance_cap(word: str) -> int:
    """The most edits a word of this length may move and still be the same word."""
    return MAX_DISTANCE_SHORT if len(word) < LONG_WORD else MAX_DISTANCE_LONG


def words(text: str) -> list[str]:
    """Split a name into the words this index holds.

    Returns nothing for any non-ASCII string rather than splitting around it
    (`aapajärvensis` must not become `aapaj` + `rvensis`). Go applies the same
    test.
    """
    if not text.isascii():
        return []
    return _WORD.findall(text.lower())


def fold_ph(word: str) -> str:
    """`ph`/`f` are the same sound, folded for both the key and the distance.

    Must be in both: folding it into the key alone puts `elefant` in `elephant`'s
    bucket, then the distance charges two edits over a cap of one. Not a wider
    cap — one substitution just stops being counted twice. Mirrored by `foldPH`
    in the serving binary.
    """
    return word.replace("ph", "f")


def spelling_key(word: str) -> str:
    """The phonetic key of one already-lowercased ASCII word.

    Mirrored exactly by `spellingKey` in the serving binary, and pinned to it by
    a Go test that reads sampled rows out of this table and recomputes them.
    That test is the contract; this docstring is only the reason for it.
    """
    w = fold_ph(word)
    w = w[:1] + w[1:].replace("h", "")
    if not w:
        return ""
    out = [w[0]]
    for c in w[1:]:
        if c in VOWELS or c == out[-1]:
            continue
        out.append(c)
    return "".join(out)


def damerau(a: str, b: str, cap: int) -> int:
    """Optimal string alignment distance, abandoning once it exceeds `cap`.

    Transpositions count as one edit: `betual` -> `betula` is one swap, and
    under plain Levenshtein it is two, tying with `betual` -> `betel` (a
    different plant) where the shorter string would win.
    """
    la, lb = len(a), len(b)
    if abs(la - lb) > cap:
        return cap + 1
    prev2: list[int] = []
    prev = list(range(lb + 1))
    for i in range(1, la + 1):
        cur = [i] + [0] * lb
        for j in range(1, lb + 1):
            cur[j] = min(
                prev[j] + 1, cur[j - 1] + 1, prev[j - 1] + (a[i - 1] != b[j - 1])
            )
            if i > 1 and j > 1 and a[i - 1] == b[j - 2] and a[i - 2] == b[j - 1]:
                cur[j] = min(cur[j], prev2[j - 2] + 1)
        if min(cur) > cap:
            return cap + 1
        prev2, prev = prev, cur
    return prev[lb]


# --------------------------------------------------------------------------
# Building
# --------------------------------------------------------------------------


def _table_exists(con: sqlite3.Connection, name: str) -> bool:
    return (
        con.execute(
            "SELECT count(*) FROM sqlite_master WHERE type IN ('table','view') "
            "AND name = ?",
            (name,),
        ).fetchone()[0]
        > 0
    )


def corpus_names(con: sqlite3.Connection) -> Iterator[str]:
    """Every distinct string `/v1/search` can match, from both catalogues.

    Both `search_name` and `fossil`, so `triceratopps` can reach *Triceratops*
    (a PBDB taxon, not a node).
    """
    seen: set[str] = set()
    sources = ["SELECT DISTINCT name FROM search_name"]
    if _table_exists(con, "fossil"):
        sources.append("SELECT DISTINCT name FROM fossil WHERE trim(name) <> ''")
    for sql in sources:
        for (name,) in con.execute(sql):
            if not name or name in seen:
                continue
            seen.add(name)
            yield name


def word_counts(names: Iterable[str]) -> dict[str, int]:
    """Distinct-name frequency per indexable word."""
    counts: dict[str, int] = {}
    for name in names:
        for w in set(words(name)):
            if len(w) >= MIN_INDEXED:
                counts[w] = counts.get(w, 0) + 1
    return counts


def build(con: sqlite3.Connection, log: Log = print) -> dict[str, int]:
    t0 = time.monotonic()
    con.executescript(SCHEMA)
    counts = word_counts(corpus_names(con))
    con.executemany(
        "INSERT INTO spelling VALUES (?,?,?)",
        ((spelling_key(w), w, n) for w, n in counts.items()),
    )
    con.commit()
    con.executescript(INDEXES)
    con.commit()
    log(f"  spelling: {len(counts):,} words in {time.monotonic() - t0:,.1f}s")
    return counts


# --------------------------------------------------------------------------
# Querying — the reference implementation of what the serving binary does
# --------------------------------------------------------------------------


def correct_word(con: sqlite3.Connection, word: str) -> str | None:
    """The best spelling of one word, or None to leave it exactly as typed."""
    if len(word) < MIN_CORRECTED:
        return None
    cap = distance_cap(word)
    rows = con.execute(
        "SELECT word, n FROM spelling WHERE key = ?", (spelling_key(word),)
    ).fetchall()
    # Both sides folded, since the bucket was built on the folded form.
    folded = fold_ph(word)
    best: tuple[int, int, int, str] | None = None
    for candidate, n in rows:
        # A word the corpus holds is registered, not a typo.
        if candidate == word:
            return None
        d = damerau(folded, fold_ph(candidate), cap)
        if d > cap:
            continue
        # Distance, then more widely used, then shorter, then lexicographic —
        # so the answer never depends on row order.
        scored = (d, -n, len(candidate), candidate)
        if best is None or scored < best:
            best = scored
    return best[3] if best is not None else None


def correct(con: sqlite3.Connection, text: str) -> str | None:
    """A better spelling of the whole query, or None if nothing was misspelled.

    Word by word, since with typeahead the misspelling that matters is in the
    leading token (`betual` kills the query before `pendula` is typed).
    """
    parts = words(text)
    if not parts:
        return None
    out = [correct_word(con, w) or w for w in parts]
    return " ".join(out) if out != parts else None


# --------------------------------------------------------------------------
# Gates
# --------------------------------------------------------------------------

# Real misspellings, and what they must reach. The first two were typed at
# concestor.com and pulled from Workers Logs; the rest are the ordinary English
# misspellings of animals a curious reader is most likely to make.
CORRECTIONS: tuple[tuple[str, str], ...] = (
    ("ardvark", "aardvark"),
    # The `ph`/`f` pair the key was built for, and which nothing reached until
    # the distance folded it too. See fold_ph.
    ("elefant", "elephant"),
    ("dolfin", "dolphin"),
    ("betual", "betula"),
    ("betual pendula", "betula pendula"),
    ("rinoceros", "rhinoceros"),
    ("gorila", "gorilla"),
    ("cheeta", "cheetah"),
    ("mosquitto", "mosquito"),
    ("aligator", "alligator"),
    ("pengiun", "penguin"),
    ("triceratopps", "triceratops"),
)

# Strings that must come back **uncorrected**, and each is a different way the
# feature could go wrong rather than ten of the same test.
#
#   hard maple  a real common name the corpus lacks — the whole point of the
#               issue, and no threshold here may reach it
#   hard oak    the same shape, typed by the same reader minutes later
#   zzzqqq      this project's own benchmark string, which is in the log corpus
#   suag, abot  short-word false positives, refused by MIN_CORRECTED
#   about       a *command*, not a taxon; the palette answers it client-side
#   pleasy      a six-letter non-word with 75 candidates in its bucket, all of
#               them further than one edit away — the cap doing its job
#   dog, whale  ordinary queries that already work and must not be touched
REFUSALS: tuple[str, ...] = (
    "hard maple",
    "hard oak",
    "zzzqqq",
    "suag",
    "abot",
    "about",
    "pleasy",
    "amt",
    "dog",
    "whale",
)
